python/tests/test_sklearn_vw.py

```diff
@@ -25,41 +27,52 @@ def data():
     return Dataset(x=x, y=y)
 
 
-class TestVW:
-
-    def test_validate_vw_estimator(self):
-        """
-        Run VW and VWClassifier through the sklearn estimator validation check
-
-        Note: the VW estimators fail sklearn's estimator validation check. The validator creates a new
-        instance of the estimator with the estimator's default args, '--quiet' in VW's case. At some point
-        in the validation sequence it calls fit() with some fake data.  The data gets formatted  via tovw() to:
+def test_tovw():
+    x = np.array([[1.2, 3.4, 5.6, 1.0, 10], [7.8, 9.10, 11, 0, 20]])
+    y = np.array([1, -1])
+    w = [1, 2]
 
-        2 1 | 0:0.5488135039273248 1:0.7151893663724195 2:0.6027633760716439 3:0.5448831829968969 4:0.4236547993389047 5:0.6458941130666561 6:0.4375872112626925 7:0.8917730007820798 8:0.9636627605010293 9:0.3834415188257777
+    expected = ['1 1 | 0:1.2 1:3.4 2:5.6 3:1 4:10',
+                '-1 2 | 0:7.8 1:9.1 2:11 4:20']
 
-        This gets passed into vw.learn and the python process dies with the error, "Process finished with exit code 139"
+    assert tovw(x=x, y=y, sample_weight=w) == expected
+    assert tovw(x=csr_matrix(x), y=y, sample_weight=w) == expected
 
-        At some point it would probably be worth while figuring out the problem  this and getting the two estimators to
-        pass sklearn's validation check
-        """
 
-        # check_estimator(VW)
-        # check_estimator(VWClassifier)
+class TestVW:
+    def test_check_estimator(self):
+        # run VW through the sklearn estimator validation check
+        check_estimator(VW)
 
```

Comment:
I believe what we should do here is that create a ABC from which all tests need to be derived. This way we can ensure that all tests implement the `test_check_estimator` and we can have more automation. THis will also make sure that in future we can enforce more such tests such as `repr_test` which will ensure two things:

- `__repr__`  methods are written.
- Tests for the same are done.

What do you think? @gramhagen

Reply:
hehe, yes sorry, thank _you_ for the test implementation!

i'm okay with having a base test class. it would be preferable to have the check_estimator and repr tests concretely defined in that class so no one needs to write them unless they are changing the behavior of the subclass somehow. is that what you're thinking?

Reply:
Yes. That is exactly how I see it happening.

Even if they are changing the behaviour, they can override the function but cannot omit it in their class.

Reply:
how's this look?